cocotb/bfmgen.py

```diff
@@ -0,0 +1,364 @@
+        exp = info.export_info[i]
+        bfm_export_tasks += "    task " + exp.T.__name__ 
+        
+        if len(exp.signature) > 0:
+            bfm_export_tasks += "("
+            for j in range(len(exp.signature)):
+                p = exp.signature[j]
+                bfm_export_tasks += p.ptype.vl_type() + " " + p.pname
+                if j+1 < len(exp.signature):
+                    bfm_export_tasks += ", "
+            bfm_export_tasks += ");\n"
+        else:
+            bfm_export_tasks += ";\n"
+        bfm_export_tasks += "    begin\n"
+        bfm_export_tasks += "        $cocotb_bfm_begin_msg(bfm_id, " + str(i) + ");\n"
+        for p in exp.signature:
+            if p.ptype.s:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_si(bfm_id, " + p.pname + ");\n"
+            else:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_ui(bfm_id, " + p.pname + ");\n"
+            
+        bfm_export_tasks += "        $cocotb_bfm_end_msg(bfm_id);\n"
+        bfm_export_tasks += "    end\n"
+        bfm_export_tasks += "    endtask\n"
+        
+    
+    impl_param_m = {
+        "bfm_classname" : info.T.__module__ + "." + info.T.__qualname__,
+        "bfm_import_calls" : bfm_import_calls,
+        "bfm_export_tasks" : bfm_export_tasks
+        }
+    
+    cocotb_bfm_api_impl = '''
+    reg signed[31:0]      bfm_id;
+    event                 bfm_ev;
+    reg signed[31:0]      bfm_msg_id;
+    
+${bfm_export_tasks}
+    
+    initial begin
+      bfm_id = $cocotb_bfm_register("${bfm_classname}", bfm_ev);
+      
+      while (1) begin
+          bfm_msg_id = $cocotb_bfm_claim_msg(bfm_id);
+          
+          case (bfm_msg_id)
+${bfm_import_calls}
+              -1: begin
+                  @(bfm_ev);
+              end
+          endcase
+          
+      end
+    end
+    '''
+   
+    param_m = {
+        "cocotb_bfm_api_impl" : Template(cocotb_bfm_api_impl).safe_substitute(impl_param_m)
+        }
+    
+    
+    return t.safe_substitute(param_m)
+
+def bfm_generate_vl(args):
+    inst = BfmMgr.inst()
+    
+    with open(args.o, "w") as out:
+        out.write("//***************************************************************************\n")
+        out.write("//* BFMs file for cocotb. \n")
+        out.write("//* Note: This file is generated by cocotb-bfmgen. Do Not Edit\n")
+        out.write("//***************************************************************************\n")
+
+        for t in inst.bfm_type_info_m.keys():
+            info = inst.bfm_type_info_m[t]
+        
+            if cocotb.bfm_vlog not in info.hdl.keys():
+                raise Exception("BFM \"" + t.__name__ + "\" does not support Verilog")
+        
+            with open(info.hdl[cocotb.bfm_vlog], "r") as template_f:
+                template = template_f.read()
+        
+            out.write(process_template_vl(template, info))
+        
+def process_template_sv(template, bfm_name, info):
```

Comment:
Please add a docstring here that explains the arguments and return value

Reply:
Done